# Task 1: Data Preparation

### Preliminaries
Import necessary modules.

In [2]:
# Module imports
import warnings
import pandas as pd
import numpy as np
# Disable warnings and allow large columns
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)

### Import data
Load data into notebook for subsequent investigation.

In [15]:
df = pd.read_csv('https://raw.githubusercontent.com/floeck/nba-analysis/master/NBA_players_stats.csv', sep = ',')

Take a look to see it has imported correctly.

In [22]:
df.sample(n=5, random_state=9)

Rk             Player Pos  Age   Tm   G  GS   MP   FG  FGA    FG%  3P  \
249  244        Cory Joseph  PG   29  SAC  36   2  749   90  201  0.448  24   
411  396  Mitchell Robinson   C   22  NYK  27  27  778  103  156  0.660   0   
452  433     Garrett Temple  SG   34  CHI  33  12  936  107  250  0.428  45   
200  197  Willy Hernangómez   C   26  NOP  20   2  359   60  110  0.545   1   
340  331    Larry Nance Jr.  PF   28  CLE  19  18  635   70  145  0.483  24   

     3PA    3P%   2P  2PA    2P%  FT  FTA    FT%  ORB  DRB  TRB  AST  STL  \
249   75  0.320   66  126  0.524  27   35  0.771   20   58   78   85   24   
411    0    NaN  103  156  0.660  22   46  0.478   97  129  226   14   32   
452  131  0.344   62  119  0.521  31   38  0.816   20   87  107   68   31   
200    5  0.200   59  105  0.562  24   36  0.667   53   86  139   26    7   
340   63  0.381   46   82  0.561  12   21  0.571   27  102  129   61   37   

     BLK  TOV  PF  PTS  
249    7   36  83  231  
411   41   22  78  228  
452   18   33  77  290  
200   10   11  27  145  
340   10   28  41  176

Looks imported correctly, so now we're going to go ahead and examine the data frame.

### Data examination & cleaning

Now to examine the shape of the dataframe and detect the presence of any null values. We will also look at summary statistics of the features.

In [20]:
print(f"Shape of the dataset is {df.shape} \n")
print(df.dtypes)

Shape of the dataset is (512, 29) 

Rk          int64
Player     object
Pos        object
Age         int64
Tm         object
G           int64
GS          int64
MP          int64
FG          int64
FGA         int64
FG%       float64
3P          int64
3PA         int64
3P%       float64
2P          int64
2PA         int64
2P%       float64
FT          int64
FTA         int64
FT%       float64
ORB         int64
DRB         int64
TRB         int64
AST         int64
STL         int64
BLK         int64
TOV         int64
PF          int64
PTS         int64
dtype: object


In [21]:
df.isnull().sum()

Rk         0
Player     0
Pos        0
Age        0
Tm         0
G          0
GS         0
MP         0
FG         0
FGA        0
FG%        3
3P         0
3PA        0
3P%       33
2P         0
2PA        0
2P%        7
FT         0
FTA        0
FT%       32
ORB        0
DRB        0
TRB        0
AST        0
STL        0
BLK        0
TOV        0
PF         0
PTS        0
dtype: int64

In [25]:
df.describe(include='object')

Player  Pos   Tm
count            512  512  512
unique           492   20   55
top     James Harden   SG  BRK
freq               3  117   19

In [27]:
df.describe(include='float64')

FG%         3P%         2P%         FT%
count  509.000000  479.000000  505.000000  480.000000
mean     0.439961    0.316823    0.501632    0.742929
std      0.120485    0.140463    0.143022    0.171721
min      0.000000    0.000000    0.000000    0.000000
25%      0.391000    0.285500    0.444000    0.667000
50%      0.442000    0.349000    0.510000    0.772000
75%      0.500000    0.394000    0.575000    0.851000
max      1.000000    1.000000    1.000000    1.000000

In [28]:
df.describe(include='int64')

Rk         Age           G          GS           MP  \
count  512.000000  512.000000  512.000000  512.000000   512.000000   
mean   248.046875   26.025391   22.515625   10.603516   512.730469   
std    141.539473   12.099932   10.762502   12.697945   374.173372   
min      1.000000  -19.000000    1.000000    0.000000     4.000000   
25%    126.750000   22.000000   14.000000    0.000000   147.750000   
50%    250.500000   25.000000   25.000000    3.000000   497.000000   
75%    370.250000   28.000000   33.000000   20.000000   814.250000   
max    492.000000  280.000000   38.000000   37.000000  1358.000000   

               FG         FGA          3P         3PA          2P         2PA  \
count  512.000000  512.000000  512.000000  512.000000  512.000000  512.000000   
mean    87.326172  187.789062   27.218750   73.910156   60.107422  113.878906   
std     82.210894  170.428063   29.654749   75.700355   63.802427  116.742818   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     18.000000   43.000000    3.000000    9.000000   10.000000   22.000000   
50%     64.500000  145.000000   17.000000   50.500000   39.000000   75.000000   
75%    131.250000  275.000000   45.000000  119.250000   88.250000  171.000000   
max    379.000000  764.000000  169.000000  411.000000  329.000000  543.000000   

               FT         FTA         ORB         DRB         TRB         AST  \
count  512.000000  512.000000  512.000000  512.000000  512.000000  512.000000   
mean    36.658203   47.070312   20.808594   73.439453   94.248047   52.925781   
std     45.406517   56.053013   24.399771   66.360902   86.459582   60.983960   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      6.000000    8.000000    4.000000   18.750000   23.000000    9.000000   
50%     21.000000   29.000000   13.000000   57.500000   73.000000   35.000000   
75%     48.000000   63.250000   27.000000  110.000000  140.000000   69.000000   
max    298.000000  363.000000  159.000000  357.000000  472.000000  345.000000   

              STL         BLK         TOV          PF           PTS  
count  512.000000  512.000000  512.000000  512.000000    512.000000  
mean    16.050781   10.509766   28.810547   41.898438    333.275391  
std     13.501952   13.058198   28.076169   31.061290   1550.220911  
min      0.000000    0.000000    0.000000    0.000000      0.000000  
25%      4.000000    2.000000    7.000000   14.000000     50.000000  
50%     14.000000    6.000000   22.000000   39.500000    180.000000  
75%     25.000000   14.000000   40.250000   64.000000    356.750000  
max     63.000000  111.000000  149.000000  228.000000  28800.000000

Here is a summary of the information gathered above.

* The shape (512, 92) matches the assignment spec.

* There are null values detected in the columns of `FG%`, `3P%`, `2P%` and `FT%`. 

* In `objects` description, although the feature players looks suitable, `Pos` and `Tm` do not. As per the spec, there should only be 7 unique positions, whereas 20 are detected. Additonally, there should only be 31 teams, whereas 55 are detected.

* In `float64` description, counts and values look ok due to presence of null values detected previously.

* In `int64` description, there are a few curious values. Min age is -19, and max is 280, both of which are illogical. Additionally, looking at `max` of each, something else comes up. Max of `PF` is 228, which is unlikely given that would mean someone had fouled a maximum 6 times over every 38 games (6*38 = 228). Also, max of `PTS` is 28800, which is over the max of 2000 as per the spec (..and logic).


* The min age of players is -19, and max is 280, which is illogical.


In [31]:
# Clean the data
teams = ['MIA', 'MIL', 'NOP', 'SAS', 'PHO', 'MEM', 'TOT', 'BRK', 'CLE', 'ORL', 'LAL', 'POR', 'TOR', 'CHI', 'WAS', 'UTA', 'SAC', 'CHO', 'NYK', 'DEN', 'LAC', 'GSW', 'OKC', 'MIN', 'DET', 'DAL', 'IND', 'ATL', 'PHI', 'BOS', 'HOU']
positions = ['PF', 'PG', 'C', 'SG', 'SF', 'PG-SG', 'SF-PF']

In [30]:
df.to_csv('cleaned_NBA_players_stats.csv', index=False)

# Task 2: Data Exploration

## Task 2.1 
Explore the players' total points: Please analyze the composition of the total points of the top five players with the most points.

In [ ]:
# Code goes after this line by adding cells


## Task 2.2 
Assuming that the data collector makes an entry error when collecting data, it can be ensured that the error occurred in the 3P, 3PA and 3P% columns, but it is not sure which player's information the error lies on. Please try to explore the error by visualization to identify how many errors there are and try to fix it.


In [ ]:
# Code goes after this line by adding cells


## Task 2.3 
Please analyze the relationship between the player's total points and the rest features (columns). Please use at least three other columns.


In [29]:
# Code goes after this line by adding cells
df.corr(method='spearman')

Rk       Age         G        GS        MP        FG       FGA  \
Rk   1.000000 -0.025736 -0.058665 -0.073919 -0.058704 -0.042869 -0.041182   
Age -0.025736  1.000000  0.211046  0.235098  0.254296  0.235297  0.235538   
G   -0.058665  0.211046  1.000000  0.646779  0.892085  0.820286  0.818044   
GS  -0.073919  0.235098  0.646779  1.000000  0.838896  0.804722  0.800294   
MP  -0.058704  0.254296  0.892085  0.838896  1.000000  0.956823  0.957656   
FG  -0.042869  0.235297  0.820286  0.804722  0.956823  1.000000  0.989092   
FGA -0.041182  0.235538  0.818044  0.800294  0.957656  0.989092  1.000000   
FG% -0.041641  0.086634  0.296663  0.301045  0.328799  0.416260  0.297593   
3P  -0.033817  0.249994  0.657724  0.610883  0.765731  0.750832  0.808104   
3PA -0.024668  0.245656  0.669796  0.623465  0.779545  0.760617  0.824525   
3P% -0.015127  0.157331  0.221392  0.217764  0.270529  0.284226  0.286331   
2P  -0.031083  0.192710  0.761974  0.767841  0.894932  0.961354  0.928662   
2PA -0.025877  0.191151  0.763834  0.772153  0.902947  0.964981  0.947174   
2P% -0.011532  0.073318  0.210734  0.181397  0.213845  0.263431  0.163168   
FT  -0.049495  0.227811  0.728512  0.755797  0.874162  0.921898  0.910947   
FTA -0.049085  0.215865  0.733142  0.764316  0.873783  0.921168  0.902797   
FT% -0.000862  0.166031  0.148403  0.103087  0.215014  0.229925  0.269971   
ORB -0.061588  0.169850  0.721925  0.655145  0.759391  0.743591  0.694198   
DRB -0.073015  0.247201  0.823301  0.784936  0.912190  0.897540  0.873924   
TRB -0.070758  0.233963  0.818652  0.770520  0.895631  0.880344  0.847958   
AST -0.037472  0.305442  0.739162  0.737407  0.886228  0.879445  0.895767   
STL -0.019495  0.253925  0.789995  0.735140  0.898916  0.859383  0.868907   
BLK -0.026091  0.151144  0.621862  0.609741  0.688493  0.680662  0.637264   
TOV -0.020275  0.234701  0.792014  0.783616  0.922189  0.931424  0.929020   
PF  -0.016473  0.198073  0.868279  0.750188  0.901741  0.857073  0.840622   
PTS -0.064446  0.245981  0.806848  0.795295  0.945236  0.984668  0.981199   

          FG%        3P       3PA       3P%        2P       2PA       2P%  \
Rk  -0.041641 -0.033817 -0.024668 -0.015127 -0.031083 -0.025877 -0.011532   
Age  0.086634  0.249994  0.245656  0.157331  0.192710  0.191151  0.073318   
G    0.296663  0.657724  0.669796  0.221392  0.761974  0.763834  0.210734   
GS   0.301045  0.610883  0.623465  0.217764  0.767841  0.772153  0.181397   
MP   0.328799  0.765731  0.779545  0.270529  0.894932  0.902947  0.213845   
FG   0.416260  0.750832  0.760617  0.284226  0.961354  0.964981  0.263431   
FGA  0.297593  0.808104  0.824525  0.286331  0.928662  0.947174  0.163168   
FG%  1.000000  0.010235 -0.031881  0.199141  0.522078  0.434011  0.786804   
3P   0.010235  1.000000  0.989232  0.578311  0.578392  0.613192 -0.029960   
3PA -0.031881  0.989232  1.000000  0.474294  0.595546  0.631257 -0.028416   
3P%  0.199141  0.578311  0.474294  1.000000  0.145669  0.163087 -0.076250   
2P   0.522078  0.578392  0.595546  0.145669  1.000000  0.990808  0.348967   
2PA  0.434011  0.613192  0.631257  0.163087  0.990808  1.000000  0.232871   
2P%  0.786804 -0.029960 -0.028416 -0.076250  0.348967  0.232871  1.000000   
FT   0.389451  0.651754  0.661871  0.228258  0.911607  0.918012  0.225939   
FTA  0.427114  0.611670  0.622700  0.199214  0.925657  0.927420  0.258876   
FT% -0.139483  0.415511  0.397979  0.327033  0.124030  0.156550 -0.169879   
ORB  0.538483  0.352570  0.362821  0.058275  0.796457  0.772894  0.373539   
DRB  0.431531  0.606500  0.617054  0.200556  0.882085  0.875265  0.299603   
TRB  0.475298  0.548572  0.559453  0.161821  0.884037  0.871755  0.333377   
AST  0.218675  0.733497  0.747734  0.262220  0.835250  0.861422  0.084786   
STL  0.256539  0.703218  0.721181  0.234550  0.817682  0.831956  0.152574   
BLK  0.479711  0.345979  0.352366  0.075278  0.722965  0.697544  0.354395   
TOV  0.329790  0.679928  0.696665  0.213800  0.910000  0.923083  0.179054   
PF   0